server/routerlicious/packages/lambdas/src/nexus/utils.ts

```diff
@@ -65,3 +65,18 @@ export const isWriter = (scopes: string[], mode: ConnectionMode) =>
 	hasWriteAccess(scopes) && mode === "write";
 
 export const getRoomId = (room: IRoom) => `${room.tenantId}/${room.documentId}`;
+export const getClientSpecificRoomId = (clientId: string) => `client#${clientId}`;
+
+export function isSentSignalMessage(obj: any): obj is ISentSignalMessage {
```

Comment:
Not `obj: unknown` or since parameter has been named `obj`, `obj: object`?

Reply:
I think 'any' might be justified here, I'm open to other approaches but I don't see how 'unknown' or 'obj' doesn't eventually lead to doing 'as any' or 'as ISentSignalMessage'.

Reply:
Unless you would prefer to do something like this? 

```typescript


export function isSentSignalMessage(obj: unknown): obj is ISentSignalMessage {
    return (
        typeof obj === "object" &&
        obj !== null &&
        "content" in obj &&
        (typeof (obj as ISentSignalMessage).type === "string" || (obj as ISentSignalMessage).type === undefined) &&
        (typeof (obj as ISentSignalMessage).clientConnectionNumber === "number" ||
            (obj as ISentSignalMessage).clientConnectionNumber === undefined) &&
        (typeof (obj as ISentSignalMessage).referenceSequenceNumber === "number" ||
            (obj as ISentSignalMessage).referenceSequenceNumber === undefined) &&
        (typeof (obj as ISentSignalMessage).targetClientId === "string" || (obj as ISentSignalMessage).targetClientId === undefined)
    );
}
```

Reply:
Really should very rarely use `any` and `as` should also be avoided. (Pretty much only use `as` when typescript is missing a feature or has a bug.)
`unknown` can be use and we don't need `as`. This should work:
```typescript
function isSentSignalMessage(obj: unknown): obj is ISentSignalMessage {
	return (
		typeof obj === "object" &&
		obj !== null &&
		"content" in obj &&
		(!("type" in obj) || typeof obj.type === "string") &&
		(!("clientConnectionNumber" in obj) || typeof obj.clientConnectionNumber === "number") &&
		(!("referenceSequenceNumber" in obj) || typeof obj.referenceSequenceNumber === "number") &&
		(!("targetClientId" in obj) || typeof obj.targetClientId === "string")
	);
}
```

Reply:
typescript compiler doesn't seem to like this:

<img width="997" alt="Screenshot 2024-04-24 at 4 01 46 PM" src="https://github.com/microsoft/FluidFramework/assets/143546745/316340db-ba74-4ff3-9ed4-fa2da99fb708">

Reply:
Oh my. That likely means server is using an older version of typescript that doesn't make it as easy to write. Looks like 4.5.5; not sure when `"foo" in bar` assertion inference was added.
I can say the logic with `any` is not the correct checks. What I provided is the proper ecmascript.
If we can't get server/routerlicious updated to a recent version of typescript (recommend 5.4.5+ but at least 5.1.6 like client group), then helper function can be used make typescript see the light.

Reply:
Looks like the fix/feature is in 4.9.5 and that allows build to pass. 5.4.5 will yield failures.
I noticed that when updating to 4.9.5 that existing tooling requires peer of typescript 5.0.0+:
```
└─┬ @fluid-tools/build-cli 0.34.0
  └─┬ @fluid-tools/version-tools 0.34.0
    └─┬ @oclif/core 3.14.1
      └─┬ tsconfck 3.0.0
        └── ✕ unmet peer typescript@^5.0.0: found 4.9.5
```
@tylerbutler, are the build tools working properly for `server/routerlicious` without being on typescript 5+?
Using 5.4.5 has type-test complaints (perhaps legit) and at least one other break.

Reply:
If we can't get to at least 4.9.5, the helper and use looks like:
```typescript
// assertion helper for typescript before 4.9.5
function hasProperty<TValue extends object, TProperty extends string>(
	obj: TValue,
	prop: TProperty,
): obj is TValue & Record<TProperty, unknown>
{
	return prop in obj;
}

function isSentSignalMessage(obj: unknown): obj is ISentSignalMessage {
	return (
		typeof obj === "object" &&
		obj !== null &&
		"content" in obj &&
		(!hasProperty(obj, "type") || typeof obj.type === "string") &&
		(!hasProperty(obj, "clientConnectionNumber") ||	typeof obj.clientConnectionNumber === "number") &&
		(!hasProperty(obj, "referenceSequenceNumber") || typeof obj.referenceSequenceNumber === "number") &&
		(!hasProperty(obj, "targetClientId") || typeof obj.targetClientId === "string")
	);
}
```

Reply:
I opted for this

Reply:
> @tylerbutler, are the build tools working properly for `server/routerlicious` without being on typescript 5+?
> Using 5.4.5 has type-test complaints (perhaps legit) and at least one other break.

Yes they are, but newer build-tools versions don't have the peer-dependency incompatibility as far as I can tell. But regardless bumping both build tools and typescript in server seems the right thing to do. PRs:

https://github.com/microsoft/FluidFramework/pull/20848
https://github.com/microsoft/FluidFramework/pull/20849